upb/decode_fast.c

```diff
@@ -0,0 +1,479 @@
+
+#include "upb/decode.h"
+
+#include "upb/port_def.inc"
+
+#define UPB_PARSE_PARAMS                                                      \
+  upb_decstate *d, const char *ptr, upb_msg *msg, const upb_msglayout *table, \
+      uint64_t hasbits, uint64_t data
+
+#define UPB_PARSE_ARGS d, ptr, msg, table, hasbits, data
+
+#define RETURN_GENERIC(msg)  \
+  /* fprintf(stderr, msg); */ \
+  return fastdecode_generic(UPB_PARSE_ARGS);
+
+typedef enum {
+  CARD_s = 0,  /* Singular (optional, non-repeated) */
+  CARD_o = 1,  /* Oneof */
+  CARD_r = 2   /* Repeated */
+} upb_card;
+
+UPB_FORCEINLINE
+const char *fastdecode_dispatch(upb_decstate *d, const char *ptr, upb_msg *msg,
+                                const upb_msglayout *table, uint64_t hasbits) {
+  uint16_t tag;
+  uint64_t data = 0;
+  size_t idx;
+  if (UPB_UNLIKELY(ptr >= d->fastlimit)) {
+    if (UPB_LIKELY(ptr == d->limit)) {
+      *(uint32_t*)msg |= hasbits >> 16;  /* Sync hasbits. */
+      return ptr;
+    }
+    RETURN_GENERIC("dispatch hit end\n");
+  }
+  memcpy(&tag, ptr, 2);
+  idx = (tag & 0xf8) >> 3;
+  data = table->field_data[idx] ^ tag;
+  return table->field_parser[idx](UPB_PARSE_ARGS);
+}
+
+UPB_FORCEINLINE
+static bool fastdecode_checktag(uint64_t data, int tagbytes) {
+  if (tagbytes == 1) {
+    return (data & 0xff) == 0;
+  } else {
+    return (data & 0xffff) == 0;
+  }
+}
+
+UPB_FORCEINLINE
+static uint16_t fastdecode_readtag(const char *ptr, int tagbytes) {
+  uint16_t ret = 0;
+  memcpy(&ret, ptr, tagbytes);
+  return ret;
+}
+
+UPB_FORCEINLINE
+static void *fastdecode_getfield_ofs(upb_decstate *d, const char *ptr,
+                                     upb_msg *msg, size_t ofs, uint64_t *data,
+                                     uint64_t *hasbits, upb_array **outarr,
+                                     void **end, int tagbytes, int valbytes,
+                                     upb_card card) {
+  void *field = (char *)msg + ofs;
+
+  switch (card) {
+    case CARD_s:
+      *hasbits |= *data;
+      return field;
+    case CARD_o: {
+      uint32_t *case_ptr = UPB_PTR_AT(msg, (*data >> 16) & 0xffff, uint32_t);
+      *case_ptr = (*data >> 32) & 0xffff;
+      return field;
+    }
+    case CARD_r: {
+      uint8_t elem_size_lg2 = __builtin_ctz(valbytes);
+      upb_array **arr_p = field;
+      upb_array *arr;
+      uint16_t expected_tag;
+      *hasbits >>= 16;
+      *(uint32_t*)msg |= *hasbits;
```

Comment:
Why are you syncing hasbits here?

Reply:
The idea was to free up a register by killing *hasbits. I was seeing spilling, and I thought as long as we are spilling we might as well spill it to its real destination.